csr/csr.go

```diff
@@ -252,32 +248,134 @@ func getNodeCSRV1Beta1(ctx context.Context, client kubernetes.Interface, nodeNam
 		return csrList.Items[i].GetCreationTimestamp().After(csrList.Items[j].GetCreationTimestamp().Time)
 	})
 
-	for _, item := range csrList.Items {
-		item := item
-		ok, err := isNodeCSR(item.Name, item.Spec.Request, nodeName)
+	for _, csr := range csrList.Items {
+		sn, err := getSubjectCommonName(csr.Name, csr.Spec.Request)
 		if err != nil {
 			return nil, err
 		}
-		if ok {
+		if sn == fmt.Sprintf("system:node:%s", nodeName) {
 			return &Certificate{
-				V1Beta1: &item,
+				V1Beta1: &csr,
 			}, nil
 		}
 	}
 
 	return nil, ErrNodeCertificateNotFound
 }
 
-func isNodeCSR(csrName string, csrRequest []byte, nodeName string) (bool, error) {
+func WatchAndApproveNodeCSR(ctx context.Context, log logrus.FieldLogger, client kubernetes.Interface, c chan *Certificate) error {
+	var w watch.Interface
+	var err error
+	b := waitext.DefaultExponentialBackoff()
+	err = waitext.Retry(
+		ctx,
+		b,
+		waitext.Forever,
+		func(ctx context.Context) (bool, error) {
+			w, err = client.CertificatesV1().CertificateSigningRequests().Watch(ctx, getOptions(certv1.KubeAPIServerClientKubeletSignerName))
+			if err != nil {
+				w, err = client.CertificatesV1beta1().CertificateSigningRequests().Watch(ctx, getOptions(certv1beta1.KubeAPIServerClientKubeletSignerName))
+				if err != nil {
+					return true, fmt.Errorf("fail to open v1 and v1beta watching client: %w", err)
+				}
+			}
+			return false, nil
+		},
+		func(err error) {
+			log.Warnf("retrying: %v", err)
```

Comment:
Context on what is being retried would be useful here (more details in the message)

Reply:
context is there, inside log fields
example
```
time="2024-05-21T19:21:39+02:00" level=warning msg="retrying: fail to open v1 and v1beta watching client: ups error" RunAutoApprove=auto-approve-csr
```